[CI] ChangePolicyforIndexIT.testChangePolicyForIndex

### CI Link

https://gradle-enterprise.elastic.co/s/yp56iltyyinha

### Repro line

./gradlew ':x-pack:plugin:ilm:qa:multi-node:javaRestTest' --tests "org.elasticsearch.xpack.ilm.ChangePolicyforIndexIT.testChangePolicyForIndex" -Dtests.seed=8825024EAE3089DD -Dtests.locale=en-PH -Dtests.timezone=Asia/Karachi -Druntime.java=16

### Does it reproduce?

No

### Applicable branches

7.17 (probably all 8+)

### Failure history

_No response_

### Failure excerpt

This is rather concerning
```
java.lang.AssertionError: policy step registry cache failed sanity check
	at org.elasticsearch.xpack.ilm.PolicyStepsRegistry.getStep(PolicyStepsRegistry.java:378) ~[?:?]
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.getCurrentStep(IndexLifecycleRunner.java:108) ~[?:?]
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.runPolicyAfterStateChange(IndexLifecycleRunner.java:430) ~[?:?]
	at org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(IndexLifecycleService.java:416) ~[?:?]
	at org.elasticsearch.xpack.ilm.IndexLifecycleService.clusterChanged(IndexLifecycleService.java:319) ~[?:?]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:573) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:559) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:519) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:428) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.access$000(ClusterApplierService.java:56) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:718) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:262) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:225) ~[elasticsearch-7.17.2-SNAPSHOT.jar:7.17.2-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?]
	at java.lang.Thread.run(Thread.java:831) [?:?]
[2022-03-15T09:34:09.657115416Z] [BUILD] Stopping node
```

The same problem affected the HLRC test suite in https://gradle-enterprise.elastic.co/s/tag573ac7s4ak:

```
[2022-03-15T15:26:43,701][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [integTest-0] fatal error in thread [elasticsearch[integTest-0][clusterApplierService#updateTask][T#1]], exiting
java.lang.AssertionError: policy step registry cache failed sanity check
        at org.elasticsearch.xpack.ilm.PolicyStepsRegistry.getStep(PolicyStepsRegistry.java:377) ~[?:?]
        at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.getCurrentStep(IndexLifecycleRunner.java:108) ~[?:?]
        at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.runPolicyAfterStateChange(IndexLifecycleRunner.java:431) ~[?:?]
        at org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(IndexLifecycleService.java:418) ~[?:?]
        at org.elasticsearch.xpack.ilm.IndexLifecycleService.clusterChanged(IndexLifecycleService.java:321) ~[?:?]
        at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:564) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:550) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:510) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:428) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:717) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:260) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:223) ~[elasticsearch-8.1.1-SNAPSHOT.jar:8.1.1-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
[2022-03-15T15:32:51.398205278Z] [BUILD] Stopping node
```

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
@@ -370,7 +370,8 @@ public class PolicyStepsRegistry {
             cachedSteps.put(indexMetadata.getIndex(), Tuple.tuple(indexMetadata, s));
             // assert that the cache works as expected -- that is, if we put something into the cache,
             // we should get back the same thing if we were to invoke getStep again with the same arguments
-            assert s == getCachedStep(indexMetadata, stepKey) : "policy step registry cache failed sanity check";
+            Step found = getCachedStep(indexMetadata, stepKey);
+            assert s == found : "policy step registry cache failed sanity check (expected [" + s + "], found [" + found + "])";
         }
         return s;
     }

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
@@ -368,10 +368,6 @@ public class PolicyStepsRegistry {
         final Step s = phaseSteps.stream().filter(step -> step.getKey().equals(stepKey)).findFirst().orElse(null);
         if (s != null) {
             cachedSteps.put(indexMetadata.getIndex(), Tuple.tuple(indexMetadata, s));
-            // assert that the cache works as expected -- that is, if we put something into the cache,
-            // we should get back the same thing if we were to invoke getStep again with the same arguments
-            Step found = getCachedStep(indexMetadata, stepKey);
-            assert s == found : "policy step registry cache failed sanity check (expected [" + s + "], found [" + found + "])";
         }
         return s;
     }

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
@@ -502,6 +502,9 @@ public class PolicyStepsRegistryTests extends ESTestCase {
         for (int i = 0; i < scaledRandomIntBetween(100, 1000); i++) {
             LifecycleAction action = randomValueOtherThan(MigrateAction.DISABLED, () -> randomFrom(phase.getActions().values()));
             Step step = randomFrom(action.toSteps(client, phaseName, MOCK_STEP_KEY, null));
+            // if the step's key is different from the previous iteration of the loop, then the cache will be updated, and we'll
+            // get a non-cached response. if the step's key happens to be the same as the previous iteration of the loop, then
+            // we'll get a cached response. so this loop randomly tests both cached and non-cached responses.
             Step actualStep = registry.getStep(indexMetadata, step.getKey());
             assertThat(actualStep.getKey(), equalTo(step.getKey()));
         }

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
@@ -50,6 +50,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.SortedMap;
 import java.util.TreeMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static org.elasticsearch.cluster.metadata.LifecycleExecutionState.ILM_CUSTOM_METADATA_KEY;
 import static org.hamcrest.Matchers.containsString;
@@ -464,5 +466,72 @@ public class PolicyStepsRegistryTests extends ESTestCase {
         gotStep = registry.getStep(metadata.index(index), shrinkStep.getKey());
         assertThat(((ShrinkStep) shrinkStep).getNumberOfShards(), equalTo(2));
         assertThat(((ShrinkStep) gotStep).getNumberOfShards(), equalTo(1));
+    }
+
+    public void testGetStepMultithreaded() throws Exception {
+        Client client = mock(Client.class);
+        Mockito.when(client.settings()).thenReturn(Settings.EMPTY);
+
+        LifecyclePolicy policy = LifecyclePolicyTests.randomTimeseriesLifecyclePolicyWithAllPhases("policy");
+        String phaseName = randomFrom(policy.getPhases().keySet());
+        Phase phase = policy.getPhases().get(phaseName);
+
+        LifecycleExecutionState lifecycleState = LifecycleExecutionState.builder()
+            .setPhaseDefinition(Strings.toString(new PhaseExecutionInfo(policy.getName(), phase, 1, randomNonNegativeLong())))
+            .build();
+        IndexMetadata indexMetadata = IndexMetadata.builder("test")
+            .settings(
+                Settings.builder()
+                    .put("index.number_of_shards", 1)
+                    .put("index.number_of_replicas", 0)
+                    .put("index.version.created", Version.CURRENT)
+                    .put(LifecycleSettings.LIFECYCLE_NAME, "policy")
+                    .build()
+            )
+            .putCustom(ILM_CUSTOM_METADATA_KEY, lifecycleState.asMap())
+            .build();
+
+        SortedMap<String, LifecyclePolicyMetadata> metas = new TreeMap<>();
+        metas.put("policy", new LifecyclePolicyMetadata(policy, Collections.emptyMap(), 1, randomNonNegativeLong()));
+        IndexLifecycleMetadata meta = new IndexLifecycleMetadata(metas, OperationMode.RUNNING);
+
+        PolicyStepsRegistry registry = new PolicyStepsRegistry(REGISTRY, client, null);
+        registry.update(meta);
+
+        // test a variety of getStep calls with random actions and steps
+        for (int i = 0; i < scaledRandomIntBetween(100, 1000); i++) {
+            LifecycleAction action = randomValueOtherThan(MigrateAction.DISABLED, () -> randomFrom(phase.getActions().values()));
+            Step step = randomFrom(action.toSteps(client, phaseName, MOCK_STEP_KEY, null));
+            Step actualStep = registry.getStep(indexMetadata, step.getKey());
+            assertThat(actualStep.getKey(), equalTo(step.getKey()));
+        }
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicBoolean done = new AtomicBoolean(false);
+
+        // now, in another thread, update the registry repeatedly as fast as possible.
+        // updating the registry has the side effect of clearing the cache.
+        new Thread(() -> {
+            latch.countDown(); // signal that we're starting
+            while (done.get() == false) {
+                registry.update(meta);
+            }
+        }).start();
+
+        try {
+            latch.await(); // wait until the other thread started
+
+            // and, while the cache is being repeatedly cleared,
+            // test a variety of getStep calls with random actions and steps
+            for (int i = 0; i < scaledRandomIntBetween(100, 1000); i++) {
+                LifecycleAction action = randomValueOtherThan(MigrateAction.DISABLED, () -> randomFrom(phase.getActions().values()));
+                Step step = randomFrom(action.toSteps(client, phaseName, MOCK_STEP_KEY, null));
+                Step actualStep = registry.getStep(indexMetadata, step.getKey());
+                assertThat(actualStep.getKey(), equalTo(step.getKey()));
+            }
+        } finally {
+            // tell the other thread we're finished
+            done.set(true);
+        }
     }
 }

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistryTests.java
@@ -511,12 +511,13 @@ public class PolicyStepsRegistryTests extends ESTestCase {
 
         // now, in another thread, update the registry repeatedly as fast as possible.
         // updating the registry has the side effect of clearing the cache.
-        new Thread(() -> {
+        Thread t = new Thread(() -> {
             latch.countDown(); // signal that we're starting
             while (done.get() == false) {
                 registry.update(meta);
             }
-        }).start();
+        });
+        t.start();
 
         try {
             latch.await(); // wait until the other thread started
@@ -530,8 +531,9 @@ public class PolicyStepsRegistryTests extends ESTestCase {
                 assertThat(actualStep.getKey(), equalTo(step.getKey()));
             }
         } finally {
-            // tell the other thread we're finished
+            // tell the other thread we're finished and wait for it to die
             done.set(true);
+            t.join(1000);
         }
     }
 }

